### ISSUE TITLE
Cross Compile: cannot open source file "stdlib.h" (dependency of foo.h)

### ISSUE THREAD

**Issue by gfxnvdo** [microsoft/vscode-docs#4532]
I'm working on a cross compilation project, and I have set it up according to [these instructions](https://code.visualstudio.com/docs/cpp/configure-intellisense-crosscompilation). However, I am still seeing errors (and some of them change as I navigate the source):

### "cannot open source file "stdlib.h" (dependency of foo.h)"

The `includePath` is set to something like this:
> ${workspaceFolder}/**
> /project/sysroots/x86_64-sdk-linux/usr/lib/aarch64-linux/gcc/aarch64-linux/include-fixed/**
> /project/sysroots/x86_64-sdk-linux/usr/lib/aarch64-linux/gcc/aarch64-linux/include/**
> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/**
> /project/sysroots/aarch64-linux/usr/include/*

The project directory hierarchy looks something like this:

```
.
└── src
    ├── CMakeLists.txt
    ├── foo
    │   ├── CMakeLists.txt
    │   ├── foo.h
    │   └── foo.cpp
    ├── bar
    │   └── bar.cpp
    └── include
        └── baz.h
```

When I'm editing foo.cpp, I get squiggles for the `#include "foo.h"` line. The details say "cannot open source file "stdlib.h" (dependency of "foo.h")."

However, there are two paths above which contain stdlib.h:
* /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/**
* /project/sysroots/aarch64-linux/usr/include/*

Both of these contain stdlib.h in the root of their paths. What's more, vscode can actually find them, even if intellisense can't:

Starting in foo.cpp, I followed each include with the "Go to Definition" menu item. (In at least one file, there was a second error shown, but the first was related to "stdlib.h," so I followed that.)

foo.cpp (#include "foo.h") -> foo.h (`#include <tuple>`)
-> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/tuple (`#include <array>`)
--> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/array (`#include <stdexcept>`)
---> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/stdexcept (`#include <string>`)
----> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/string (`#include <bits/basic_string.h>`)
-----> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/bits/basic_string.h (`#include <ext/string_conversions.h>`)
------> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/ext/string_conversions.h (`#include <cstdlib>`)
-------> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/cstdlib (`#include <bits/std_abs.h>`)
--------> /project/sysroots/aarch64-linux/usr/include/c++/7.3.0/bits/std_abs.h (`#include_next <stdlib.h>`)

At this point, the errors in vscode actually changed. The "stdlib.h" error disappeared, replaced by squiggles in the comments of this file (std_abs.h), saying simply "#include errors detected. Please update your includePath. Squiggles are disabled for this translation unit (.../foo.cpp)." The same error was all that showed for the original foo.cpp, as well.

In any case, I persevered and vscode had no issue finding the correct stdlib.h

---------> /project/sysroots/aarch64-linux/usr/include/stdlib.h

So, obviously, the question is what do I need to change in my C/C++ plugin configuration to get Intellisense to function correctly?

**Comment by gfxnvdo** [comment]
Is there any feedback on this? I've tried moving the paths around, but I haven't found anything that will make the plugin happy.

**Comment by gfxnvdo** [comment]
bump

**Comment by jureid** [comment]
Hi @gfxnvdo, thanks for opening this issue! I've passed this along to our engineering team to help troubleshoot. We'll get back to you shortly.

**Comment by jureid** [comment]
Hi @gfxnvdo, thanks for your patience. Here are a couple of things you can do to correct your C++ IntelliSense configuration:

- includePath shouldn't contain any system include paths. Instead, make sure your compilerPath property is set correctly. The C++ extension uses compilerPath to look for system headers.
- For cross-compilation IntelliSense, be sure to specify any cross compiler flags in the compilerArgs property. The C++ extension will use compilerArgs to query the correct system include paths for your target architecture. 

Also, are you using CMake Tools to configure and build your project? If so, make sure your configurationProvider is set to ms-vscode.cmake-tools. You can find configurationProvider under Advanced Settings in the Edit Configurations UI: 
![image](https://user-images.githubusercontent.com/44306969/120519399-6a847900-c387-11eb-99e7-bcee65b52e7c.png)

**Comment by gfxnvdo** [comment]
### tl;dr
Perhaps I found the right combination. It seems that BOTH includePath and --sysroot compiler argument together are required to alleviate the errors (squiggles). I don't think that's how it's supposed to work (bug?), but it seems to be working right now.

### Details

I've included a facsimile of my c_cpp_properties.json file below. Removing either the non-local entries in the includePath section or the --sysroot commandArg result in the error reported above.

> includePath shouldn't contain any system include paths.

The many many combinations I tried were a process. I had previously tried including the --sysroot compiler argument provided to the compiler, but that gave me the above symptoms. So I removed it and manually added the includePaths above, one at a time, to remove the errors (squiggles). But even though I provided the paths to the missing files, the errors persisted. And I figured the inability of the --sysroot extraction of paths from the compiler was the same issue, with another level of indirection. So I filed this ticket with that combination, believing that the inability of the plugin to find files in the exact path provided was a bug. (I still think this looks like a bug.)

> be sure to specify any cross compiler flags in the compilerArgs property

The compiler flags used are:
* -m64
* -march=...
* -mtune=...
* -mfpmath=...
* -msse4.2
* --sysroot=/project/sysroots/corei7-64-linux
* --sysroot=/project/sysroots/corei7-64-linux
* -D...
* -I/project/src/include
* -O2
* -pipe
* -g
* -feliminate-unused-debug-types
* -pthread
* -Wall
* -Wextra
* -Werror
* -fPIC
* -std=...
* -o ...
* -c ...

I don't think any of these are of interest to the plugin, especially with respect to include file paths, other than the ones below (--systroot & -I)

In any case, the below combination, with both includePath and the --sysroot compilerArgs appears to eliminate the errors. But the paths appear to be redundant and this is very counterintuitive.

> Also, are you using CMake Tools to configure and build your project?

Yes, this project uses CMake, but I am not using CMake Tools because it is incompatible with C/C++ Tools (https://github.com/microsoft/vscode-cpptools/issues/5266 -- closed, but not fixed).

**c_cpp_properties.json**:
```
{
    "env": {
    },
    "configurations": [
        {
            "name": "project-arm64",
            "intelliSenseMode": "linux-clang-arm64",
            "compilerPath": "/project/sysroots/x86_64-sdk-linux/usr/bin/aarch64-linux/aarch64-linux-g++",
            "includePath": [
                "${workspaceFolder}/**",
                "/project/sysroots/x86_64-sdk-linux/usr/lib/aarch64-linux/gcc/aarch64-linux/7.3.0/include-fixed/**",
                "/project/sysroots/x86_64-sdk-linux/usr/lib/aarch64-linux/gcc/aarch64-linux/7.3.0/include/**",
                "/project/2.5.3/sysroots/aarch64-linux/usr/include/c++/7.3.0/**",
                "/project/2.5.3/sysroots/aarch64-linux/usr/include/*"
            ],
            "compilerArgs": [
                "--sysroot=/project/sysroots/corei7-64-linux",
                "-I/project/src/version"
            ],
            "cppStandard": "c++17",
            "browse": {
                "limitSymbolsToIncludedHeaders": true,
                "databaseFilename": ""
            }
        }
    ],
    "version": 4
}
```

**Comment by gregvanl** [comment]
Closing as user is unblocked and contributed a nice write up.
No other action required.
